src/IECoreDelight/Renderer.cpp

```diff
@@ -905,18 +926,116 @@ class DelightObject : public IECoreScenePreview::Renderer::ObjectInterface
 			/// \todo Implement
 		}
 
-	private :
+	protected :
 
 		const DelightHandle m_transformHandle;
 		// We keep a reference to the instance and attributes so that they
 		// remain alive for at least as long as the object does.
 		ConstDelightAttributesPtr m_attributes;
 		DelightHandleSharedPtr m_instance;
 
+	private :
+
 		bool m_haveTransform;
 
 };
 
+}  // namespace
+
+//////////////////////////////////////////////////////////////////////////
+// DelightObject
+//////////////////////////////////////////////////////////////////////////
+
+namespace
+{
+
+class DelightObject : public DelightObjectBase
+{
+
+	public :
+
+		DelightObject( NSIContext_t context, const std::string &name, DelightHandleSharedPtr instance, DelightHandle::Ownership ownership )
+			:	DelightObjectBase( context, name, instance, ownership )
+		{
+		}
+
+};
+
+} // namespace
+
+namespace
+{
+
+class DelightLight : public DelightObjectBase
+{
+
+	public :
+
+		DelightLight( NSIContext_t context, const std::string &name, DelightHandleSharedPtr instance, DelightHandle::Ownership ownership )
+			: DelightObjectBase( context, name, instance, ownership ), m_usdLightType( IECoreDelight::ShaderNetworkAlgo::USDLightType::None)
+		{
+		}
+
+		bool attributes( const IECoreScenePreview::Renderer::AttributesInterface *attributes ) override
+		{
+			DelightObjectBase::attributes( attributes );
+
+			if( const ShaderNetwork *usdLightShader = m_attributes->usdLightShader() )
+			{
+				IECoreDelight::ShaderNetworkAlgo::USDLightType usdLightType =
+					IECoreDelight::ShaderNetworkAlgo::usdLightType( usdLightShader );
+
+				if( usdLightType != m_usdLightType)
+				{
+					if( m_instance )
+					{
+						NSIDisconnect(
+							m_transformHandle.context(),
+							m_instance->name(), "",
+							m_transformHandle.name(), "objects"
+						);
+					}
+
+					m_instance = IECoreDelight::ShaderNetworkAlgo::createUSDLightGeometry(
+						m_transformHandle,
+						usdLightShader,
+						usdLightType
+					);
```

Comment:
Or is this actually a problem with DelightHandle, and `operator =` should be doing a `reset()` before it does anything else?

Reply:
> I think this can probably be fixed by doing `m_instance.reset()` before `m_instance = `, to free up the old handle before trying to reuse it?

Yes, that works nicely.



> Or is this actually a problem with DelightHandle, and `operator =` should be doing a `reset()` before it does anything else?

I don't _think_ so. Wouldn't it be at least technically valid to do

```
DelightHandle a( context, "shader:metal", ownership )
DelightHandle b( context, "shader:metal", ownership )
a = DelightHandle( context, "shader:wood", ownership )
```

If `DelightHandle` called `reset()` in `operator =`, then `b` would no longer point to a valid 3Delight object.

Reply:
I'm manually releasing the `DelightHandleSharedPtr` for now in https://github.com/GafferHQ/gaffer/pull/5603/commits/a0c61d62e4bb80bef1f2ddfe3839f68389fcb36d, happy to revisit now or as a follow-up if you want to reconsider `reset()`ing within the assignment operator.

Reply:
I think my intention was that DelightHandle should act like `unique_ptr`, in which case constructing `a` and `b` from the same node in your example would be equivalent to constructing two `unique_ptrs` from the same raw pointer. With that in mind, you'd expect `b` to now be dangling...

We don't need to deal with this in this PR though - what you've done makes sense.

Reply:
Although then DelightHandleSharedPtr is like `shared_ptr<unique_ptr>`, which is just weird. Maybe DelightHandle shouldn't even be assignable? To be continued...